### PR TITLE
Exponential Backoff for wrong password login (PLFM-5355)

### PR DESCRIPTION
**Pull request by zimingd**

### CODE REVIEW

File: lib/jdomodels/src/main/resources/schema/UnsuccessfulAttemptLockout-ddl.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE IF NOT EXISTS `UNSUCCESSFUL_ATTEMPT_LOCKOUT`(
+  `ATTEMPT_KEY` varchar(256) CHARACTER SET latin1 COLLATE latin1_bin NOT NULL,
```
Comment (zimingd) [inline lib/jdomodels/src/main/resources/schema/UnsuccessfulAttemptLockout-ddl.sql]: change to bigint unsigned (w/ enum type if need generic)
Rename to userid

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulAttemptLockoutDAOImpl.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.springframework.jdbc.core.JdbcTemplate;
+
+public class UnsuccessfulAttemptLockoutDAOImpl implements UnsuccessfulAttemptLockoutDAO {
+
+	private static final String CURRENT_TIMESTAMP_MILLIS = "CURRENT_TIMESTAMP(3)";
+
+	private static final String FROM_TABLE_FILTERED_BY_KEY = " FROM " + TABLE_UNSUCCESSFUL_ATTEMPT_LOCKOUT
+			+ " WHERE " + COL_UNSUCCESSFUL_ATTEMPT_KEY + " = ?";
+
+	private static final String SELECT_EXPIRATION_AND_LOCK_ROW = "SELECT CASE WHEN " + COL_UNSUCCESSFUL_ATTEMPT_LOCKOUT_EXPIRATION_TIMESTAMP_MILLIS + " <= "+CURRENT_TIMESTAMP_MILLIS+" THEN null ELSE " + " CAST(UNIX_TIMESTAMP(" + COL_UNSUCCESSFUL_ATTEMPT_LOCKOUT_EXPIRATION_TIMESTAMP_MILLIS + ") * 1000 AS UNSIGNED) END"
```
Comment (zimingd) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulAttemptLockoutDAOImpl.java]: SELECT FOR UPDATE on credentials DAO to ensure locking.

if user does not exist. throw NotFoundException(). Verify that manager catches and throws exception stating wrong username or password combination.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulAttemptLockoutDAOImpl.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+			+ FROM_TABLE_FILTERED_BY_KEY
+			 + " FOR UPDATE";
```
Comment (zimingd) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulAttemptLockoutDAOImpl.java]: use pojo to pass along both expiration time an current database time to caller of DAO. caller can decide on expiration?

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulAttemptLockoutDAOImpl.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+	@MandatoryWriteReadCommittedTransaction
+	@Override
+	public void removeLockout(String key) {
+		jdbcTemplate.update(REMOVE_LOCKOUT, key);
+	}
+
+	@MandatoryWriteReadCommittedTransaction
+	@Override
+	public Long getUnexpiredLockoutTimestampMillis(String key) {
```
Comment (zimingd) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulAttemptLockoutDAOImpl.java]: return pair object

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulAttemptLockoutDAOImpl.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+			+ FROM_TABLE_FILTERED_BY_KEY;
+
+	private static final String REMOVE_LOCKOUT = "DELETE " + FROM_TABLE_FILTERED_BY_KEY;
+
+	private static final String CREATE_OR_INCREMENT_ATTEMPT_COUNT = "INSERT INTO " + TABLE_UNSUCCESSFUL_ATTEMPT_LOCKOUT
+			+ " (" + COL_UNSUCCESSFUL_ATTEMPT_KEY + "," + COL_UNSUCCESSFUL_ATTEMPT_COUNT+","+ COL_UNSUCCESSFUL_ATTEMPT_LOCKOUT_EXPIRATION_TIMESTAMP_MILLIS +")"
+			+ " VALUES (?, 1, "+CURRENT_TIMESTAMP_MILLIS+") ON DUPLICATE KEY UPDATE " + COL_UNSUCCESSFUL_ATTEMPT_COUNT + "=" + COL_UNSUCCESSFUL_ATTEMPT_COUNT + "+1";
+
+	private static final String UPDATE_EXPIRATION = "UPDATE " + TABLE_UNSUCCESSFUL_ATTEMPT_LOCKOUT
+			+ " SET " + COL_UNSUCCESSFUL_ATTEMPT_LOCKOUT_EXPIRATION_TIMESTAMP_MILLIS + "="+CURRENT_TIMESTAMP_MILLIS+" + INTERVAL (? * 1000) MICROSECOND" + " WHERE " + COL_UNSUCCESSFUL_ATTEMPT_KEY + "=?";
```
Comment (zimingd) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulAttemptLockoutDAOImpl.java]: if we also return database current time, the lockout logic can be responsible for setting the new expiration based on database current time

Comment (zimingd) [reply lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulAttemptLockoutDAOImpl.java]: let basicdao perform update but manually select desired columns on fetch

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java [hunk 6]
```diff
@@ -45,23 +33,12 @@
...
-		loginFailAttemptExceedLimit.setNamespace(this.getClass().getName());
-		loginFailAttemptExceedLimit.setName(LOGIN_FAIL_ATTEMPT_METRIC_NAME);
-		loginFailAttemptExceedLimit.setValue(LOGIN_FAIL_ATTEMPT_METRIC_DEFAULT_VALUE);
-		loginFailAttemptExceedLimit.setUnit(LOGIN_FAIL_ATTEMPT_METRIC_UNIT);
-		loginFailAttemptExceedLimit.setTimestamp(new Date());
-		loginFailAttemptExceedLimit.setDimension(Collections.singletonMap("UserId", ""+principalId));
-		consumer.addProfileData(loginFailAttemptExceedLimit);
-	}
+	@Autowired
+	private AuthenticationManagerUtil authUtil;
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java]: rename to userCredentialValidator

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java [hunk 7]
```diff
@@ -45,23 +33,12 @@
 	private UserGroupDAO userGroupDAO;
 	@Autowired
 	private AuthenticationReceiptDAO authReceiptDAO;
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java:35]: should use principalAliasDao to translate username to principalid. do no rely on `userManager.lookupUserForAuthentication`(maybe remove if only used here) in `AuthenticationServiceImpl`

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java [hunk 8]
```diff
@@ -166,35 +143,42 @@ public void setTermsOfUseAcceptance(Long principalId, Boolean acceptance) {
...
+		String validAuthReceipt = null;
+		if (authenticationReceipt != null && authReceiptDAO.isValidReceipt(principalId, authenticationReceipt)){
+			validAuthReceipt = authenticationReceipt;
 		}
 
-		authenticateAndThrowException(principalId, password);
-		if (lockToken != null) {
-			authenticationThrottleMemoryCountingSemaphore.releaseLock(""+principalId, lockToken);
+		//callers that have previously logged in successfully are able to bypass lockout caused by failed attempts
+		boolean correctCredentials = validAuthReceipt != null ? authUtil.checkPassword(principalId, password) : authUtil.checkPasswordWithLock(principalId, password);
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java:152]: should return false even if password hash or principalId do not exist

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerUtil.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package org.sagebionetworks.repo.manager;
+
+import org.sagebionetworks.repo.transactions.RequiresNewReadCommitted;
+
+public interface AuthenticationManagerUtil {
+	boolean checkPassword(Long principalId, String password);
+
+	@RequiresNewReadCommitted
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerUtil.java]: remove. no effect here

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerUtilImpl.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+public class AuthenticationManagerUtilImpl implements AuthenticationManagerUtil {
+	public static final String LOGIN_FAIL_ATTEMPT_METRIC_UNIT = "Count";
+
+	public static final double LOGIN_FAIL_ATTEMPT_METRIC_DEFAULT_VALUE = 1.0;
+
+	public static final String LOGIN_FAIL_ATTEMPT_METRIC_NAME = "LoginFailAttemptExceedLimit";
+
+	public static final String UNSUCCESSFUL_LOGIN_ATTEMPT_KEY_PREFIX = "login-";
+
+	static final long REPORT_UNSUCCESSFUL_LOGIN_GREATER_OR_EQUAL_THRESHOLD = 11;
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerUtilImpl.java]: 2^18 milliseconds = 4.369 minutes

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/unsuccessfulattemptlockout/ExponentialBackoffAttemptReporter.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+	@MandatoryWriteReadCommittedTransaction
+	@Override
+	public void reportFailure() {
+		long numFailed = dao.incrementNumFailedAttempts(attemptKey);
+		long lockDurationMilliseconds = 1 << numFailed;
+		dao.setExpiration(attemptKey, lockDurationMilliseconds);
+	}
+
+	@Override
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/unsuccessfulattemptlockout/ExponentialBackoffAttemptReporter.java]: check if equals and hash code still being used in tests

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java (hunk 8 above)
Comment (zimingd) [reply services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java]: DBOAuthenticationDAOImplTest already tests forthis

Comment (zimingd) [reply services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java:152]: DBOAuthenticationDAOImplTest already tests for this

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/unsuccessfulattemptlockout/ExponentialBackoffAttemptReporter.java (hunk 11 above)
Comment (zimingd) [reply services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/unsuccessfulattemptlockout/ExponentialBackoffAttemptReporter.java]: yep makes test code much cleaner

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulLoginLockoutDAOImplAutowiredTest.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+		dao.createOrUpdateUnsuccessfulLoginLockoutInfo(dto);
+		assertEquals(dto, dao.getUnsuccessfulLoginLockoutInfoIfExist(userId));
+
+		//update the DTO
+		dto.withLockoutExpiration(128).withLockoutExpiration(2L);
+		dao.createOrUpdateUnsuccessfulLoginLockoutInfo(dto);
+		assertEquals(dto, dao.getUnsuccessfulLoginLockoutInfoIfExist(userId));
+
+		//delete the DTO
+		dao.deleteUnsuccessfulLoginLockoutInfo(userId);
```
Comment (zimingd) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/UnsuccessfulLoginLockoutDAOImplAutowiredTest.java:81]: delete is slower
reset values instead

File: lib/lib-test/src/main/java/org/sagebionetworks/repo/web/controller/ExceptionHandlers.java [hunk 13]
```diff
@@ -86,7 +86,8 @@ public TestEntry(int statusCode, ExceptionType... exceptions) {
 		new TestEntry(429,
 				new ExceptionType("org.sagebionetworks.repo.model.TooManyRequestsException", true)),
 		new TestEntry(423,
-				new ExceptionType("org.sagebionetworks.repo.model.LockedException", true)),
+				new ExceptionType("org.sagebionetworks.repo.model.LockedException", true),
+				new ExceptionType("org.sagebionetworks.repo.manager.loginlockout.UnsuccessfulLoginLockoutException", true)),
```
Comment (zimingd) [inline lib/lib-test/src/main/java/org/sagebionetworks/repo/web/controller/ExceptionHandlers.java:90]: maybe translate unix timestamp in error to time delta in seconds. ask jay

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserCredentialValidator.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package org.sagebionetworks.repo.manager;
+
+public interface UserCredentialValidator {
+	boolean checkPassword(Long principalId, String password);
+
+	boolean checkPasswordWithLock(Long principalId, String password);
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserCredentialValidator.java:6]: add javadoc

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserManagerImpl.java [hunk 15]
```diff
@@ -247,12 +245,9 @@ public void deletePrincipal(UserInfo adminUserInfo, Long principalId) throws Not
 	public PrincipalAlias lookupUserForAuthentication(String alias) {
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserManagerImpl.java:245]: change to lookupPrincipalForAlias. No logic in it. Callers must handle whether or not they get a Team

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/loginlockout/AttemptResultReporter.java [hunk 16]
```diff
@@ -0,0 +1,10 @@
+package org.sagebionetworks.repo.manager.loginlockout;
+
+/**
+ * Used to report success or failure on an attempt.
+ * @see UnsuccessfulLoginLockout#checkIsLockedOut(long)
+ */
+public interface AttemptResultReporter {
+	void reportSuccess();
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/loginlockout/AttemptResultReporter.java:8]: javadocs

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/loginlockout/ExponentialBackoffAttemptReporter.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+	@MandatoryWriteReadCommittedTransaction
+	@Override
+	public void reportSuccess() {
+		dao.deleteUnsuccessfulLoginLockoutInfo(lockoutInfo.getUserId());
+	}
+
+	@MandatoryWriteReadCommittedTransaction
+	@Override
+	public void reportFailure() {
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/loginlockout/ExponentialBackoffAttemptReporter.java:30]: make these calls idempotent

File: services/repository-managers/src/main/resources/private/managers-spb.xml [hunk 18]
```diff
@@ -245,9 +239,13 @@
 	<bean id="accessApprovalManager" class="org.sagebionetworks.repo.manager.AccessApprovalManagerImpl" />
 
 	<bean id="userManager" class="org.sagebionetworks.repo.manager.UserManagerImpl" />
-		
+
+	<bean id="authManagerUtil" class="org.sagebionetworks.repo.manager.UserCredentialValidatorImpl" />
```
Comment (zimingd) [inline services/repository-managers/src/main/resources/private/managers-spb.xml:243]: rename bean id

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/loginlockout/ExponentialBackoffAttemptReporterTest.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+		reporter.reportFailure();
+
+		UnsuccessfulLoginLockoutDTO expected = new UnsuccessfulLoginLockoutDTO(userId)
+			.withUnsuccessfulLoginCount(unsuccessfulLoginCount + 1) // 3 + 1 = 4
+				.withLockoutExpiration(databaseTimeStamp + 16); // 2^4 = 16
+
+		verify(mockUnsuccessfulLoginLockoutDAO).getDatabaseTimestampMillis();
+		verify(mockUnsuccessfulLoginLockoutDAO).createOrUpdateUnsuccessfulLoginLockoutInfo(expected);
+		verifyNoMoreInteractions(mockUnsuccessfulLoginLockoutDAO);
+	}
```
Comment (zimingd) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/loginlockout/ExponentialBackoffAttemptReporterTest.java:68]: add test for idempotence

**Review by john-hill — APPROVED** [review]

**Pull request merged**